text/src/autogluon/text/text_prediction/mx/models.py

```diff
@@ -0,0 +1,1505 @@
+        not_included = []
+    # consider the task specific fine-tuning layer as the last layer, following with pooler
+    # In addition, the embedding parameters have the smaller learning rate based on this setting.
+    if 'albert' in backbone_name:
+        # Skip if it is the ALBERT model.
+        return
+    if 'electra' in backbone_name:
+        # For ELECTRA, it's called all_encoder_layers
+        all_layers = model.encoder.all_encoder_layers
+    else:
+        # For other models, it's called all_layers
+        all_layers = model.encoder.all_layers
+    max_depth = len(all_layers) + 2
+    for key, value in model.collect_params().items():
+        if 'scores' in key:
+            value.lr_mult = layerwise_decay ** 0
+        if 'pool' in key:
+            value.lr_mult = layerwise_decay ** 1
+        if 'embed' in key:
+            value.lr_mult = layerwise_decay ** max_depth
+
+    for (layer_depth, layer) in enumerate(all_layers):
+        layer_params = layer.collect_params()
+        for key, value in layer_params.items():
+            for pn in not_included:
+                if pn in key:
+                    continue
+            value.lr_mult = layerwise_decay ** (max_depth - (layer_depth + 1))
+
+
+@use_np
+def freeze_layers(model, backbone_name, num_trainable_layers):
+    if 'albert' in backbone_name:
+        # Skip if it is the ALBERT model.
+        return
+    if 'electra' in backbone_name:
+        # For ELECTRA, it's called all_encoder_layers
+        all_layers = model.encoder.all_encoder_layers
+    else:
+        # For other models, it's called all_layers
+        all_layers = model.encoder.all_layers
+    if num_trainable_layers < 0:
+        return
+    assert num_trainable_layers <= len(all_layers)
+    for i in range(len(all_layers) - num_trainable_layers):
+        for p in all_layers[i].collect_params().values():
+            p.grad_req = 'null'
+    return
+
+
+def base_optimization_config():
+    """The basic optimization phase"""
+    cfg = CfgNode()
+    cfg.lr_scheduler = 'triangular'
+    cfg.optimizer = 'adamw'
+    cfg.early_stopping_patience = 20  # Stop if we cannot find a better checkpoint
+    cfg.optimizer_params = [('beta1', 0.9),
+                            ('beta2', 0.999),
+                            ('epsilon', 1e-6),
+                            ('correct_bias', False)]
+    cfg.begin_lr = 0.0
+    cfg.batch_size = 128
+    cfg.nbest = 1                   # Keep the top K performed models
+    cfg.per_device_batch_size = 16  # Per-device batch-size
+    cfg.auto_per_device_batch_size = True  # Whether to automatically determine the runnable
+                                           # per-device batch_size.
+    cfg.val_batch_size_mult = 2  # By default, we 2X the batch size for validation
+    cfg.lr = 1E-4
+    cfg.final_lr = 0.0
+    cfg.num_train_epochs = 10
+    cfg.warmup_portion = 0.1
+    cfg.layerwise_lr_decay = 0.8  # The layer_wise decay
+    cfg.wd = 0.01  # Weight Decay
```

Comment:
this is probably a good candidate for HPO. [Paper](https://arxiv.org/pdf/1803.09820.pdf)

> Weight decay (WD): This requires a grid search to determine the proper magnitude butusually  does  not  require  more  than  one  significant  figure  accuracy.   Use  your  knowl-edge of the dataset and architecture to decide which values to test.  For example, a morecomplex dataset requires less regularization so test smaller weight decay values, such as10−4,10−5,10−6,0.   A  shallow  architecture  requires  more  regularization  so  test  largerweight decay values, such as10−2,10−3,10−4.

Reply:
Mark as future work.